### PR TITLE
Minor optimization for refresh + fix for experiment setup.

### PR DESCRIPTION
**Pull request by glevitzky**
Changes how DoubleClick initializes RefreshManager. Specifically, we no longer create an instance unless we need one. Moreover, this PR factors out the initialization logic (i.e., the eligibility checks), so that an instance can be created when the refresh experiment header is present.

### CODE REVIEW

File: extensions/amp-a4a/0.1/test/test-refresh-manager.js [hunk 1]
```diff
@@ -54,9 +55,15 @@ describe('refresh-manager', () => {
     sandbox.restore();
   });
 
+  it('should get null from slot', () => {
```
Comment (jonkeller) [inline extensions/amp-a4a/0.1/test/test-refresh-manager.js]: How about "should get null refreshInterval"?

File: extensions/amp-a4a/0.1/test/test-refresh-manager.js [hunk 2]
```diff
@@ -160,3 +155,4 @@ describe('refresh-manager', () => {
 });
 
 
+
```
Comment (jonkeller) [inline extensions/amp-a4a/0.1/test/test-refresh-manager.js]: Nit: you could remove lines 157, 158

File: extensions/amp-a4a/0.1/test/test-refresh-manager.js (hunk 1 above)
Comment (glevitzky) [reply extensions/amp-a4a/0.1/test/test-refresh-manager.js]: Done.

File: extensions/amp-a4a/0.1/test/test-refresh-manager.js (hunk 2 above)
Comment (glevitzky) [reply extensions/amp-a4a/0.1/test/test-refresh-manager.js]: Done.

File: extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js [hunk 3]
```diff
@@ -769,6 +760,32 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
...
+    let refreshInterval = 0;
+    this.refreshManager_ = this.refreshManager_ ||
+        // Not compatible with SRA
+        (!this.useSra &&
+         // Not compatible with container types except carousel and sticky-ad
+         !getEnclosingContainerTypes(this.element).filter(container =>
+           container != ValidAdContainerTypes['AMP-CAROUSEL'] &&
+           container != ValidAdContainerTypes['AMP-STICKY-AD']).length &&
+         // Publisher must supply a refresh interval >= 30s
+         (refreshInterval = getPublisherSpecifiedRefreshInterval(
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js]: you could make a refreshmanager factory within refresh-manager code in which you place the interval check code where it would return null if invalid.  In fact it could take a verifying predicate and user().warn if invalid as I do not believe you're doing so right now if there is an invalid container type

File: extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js [hunk 4]
```diff
@@ -660,11 +660,10 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
       this.extensions_./*OK*/installExtensionForDoc(
           this.getAmpDoc(), 'amp-analytics');
     }
-
     const refreshInterval = Number(responseHeaders.get('amp-force-refresh'));
-    if (refreshInterval && !getPublisherSpecifiedRefreshInterval(
-        this.element, this.win, 'doubleclick')) {
+    if (refreshInterval) {
       this.element.setAttribute(DATA_ATTR_NAME, refreshInterval);
+      this.initRefreshManagerIfEligible_();
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js]: I don't think you need this here given layoutCallback would just trigger refresh naturally.  Also, why did you remove the getPublisherSpecifiedRefreshInterval?

File: extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js [hunk 5]
```diff
@@ -750,15 +749,7 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
...
-    this.refreshManager_ = this.useSra ||
-        getEnclosingContainerTypes(this.element).filter(container =>
-            container != ValidAdContainerTypes['AMP-CAROUSEL'] &&
-            container != ValidAdContainerTypes['AMP-STICKY-AD']).length
-            ? null
-            : this.refreshManager_ || new RefreshManager(this, {
-              visiblePercentageMin: 50,
-              continuousTimeMin: 1,
-            });
+    this.initRefreshManagerIfEligible_();
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js]: You're racing with the layoutPromise which I believe is the underlying issue here.  Instead, wait for super to complete.  Consider moving into onCreativeRender

File: extensions/amp-a4a/0.1/refresh-manager.js [hunk 6]
```diff
@@ -46,6 +51,8 @@ const TAG = 'AMP-AD';
  * @param {!Window} win
  * @param {string} adType
  * @return {?number}
+ *
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/refresh-manager.js]: nit: remove extra line

File: extensions/amp-a4a/0.1/refresh-manager.js [hunk 7]
```diff
@@ -151,13 +158,52 @@ const managers = {};
...
+/**
+ * Initializes a RefreshManager under the following conditions:
+ *   2. SRA is not enabled.
+ *   3. The slot is not inside an ad container (with the exception of
+ *      carousel and sticky ad).
+ *   4. The publisher has provided an appropriate refresh interval (>= 30s).
+ *
+ * @param {!./amp-a4a.AmpA4A} a4a
+ * @param {boolean} useSra
+ * @return {?RefreshManager}
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/refresh-manager.js]: comment about null if invalid

File: extensions/amp-a4a/0.1/refresh-manager.js [hunk 8]
```diff
@@ -151,13 +158,52 @@ const managers = {};
...
+ */
+export function getRefreshManagerForDoubleclickIfEligible(a4a, useSra) {
+  const refreshInterval = getPublisherSpecifiedRefreshInterval(
+      a4a.element, a4a.win, 'doubleclick');
+  if (!refreshInterval) {
+    return null;
+  }
+
+  if (useSra) {
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/refresh-manager.js]: Might as well move this above refreshInterval check given it is the most lightweight

File: extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js [hunk 9]
```diff
@@ -750,15 +748,6 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
     if (this.useSra && this.element.getAttribute(DATA_ATTR_NAME)) {
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js]: Remove and instead just do return super.layoutCallback();

File: extensions/amp-a4a/0.1/refresh-manager.js [hunk 10]
```diff
@@ -46,6 +51,8 @@ const TAG = 'AMP-AD';
  * @param {!Window} win
  * @param {string} adType
  * @return {?number}
+ *
+ * @visibleForTesting
  */
 export function getPublisherSpecifiedRefreshInterval(element, win, adType) {
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/refresh-manager.js]: Can't you get the adType from the element (element.getAttribute('type'))?

File: extensions/amp-a4a/0.1/refresh-manager.js [hunk 11]
```diff
@@ -151,13 +158,52 @@ const managers = {};
...
+ *   2. SRA is not enabled.
+ *   3. The slot is not inside an ad container (with the exception of
+ *      carousel and sticky ad).
+ *   4. The publisher has provided an appropriate refresh interval (>= 30s).
+ *
+ * @param {!./amp-a4a.AmpA4A} a4a
+ * @param {boolean} useSra
+ * @return {?RefreshManager}
+ */
+export function getRefreshManagerForDoubleclickIfEligible(a4a, useSra) {
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/refresh-manager.js]: Not crazy about having doubleclick specific requirements within refresh manager.  Anyway we could pass the predicate?

export function getRefreshManager(a4a, opt_predicate) {
   if (getRefreshIntervalFromPageOrElement == invalid) {
      user().warn(...);
      return null;
   }
   if (opt_predicate && !opt_predicate()) {
      user().warn(...);
      return null;
   }
   return new Refresh...
}

File: extensions/amp-a4a/0.1/refresh-manager.js (hunk 10 above)
Comment (glevitzky) [reply extensions/amp-a4a/0.1/refresh-manager.js]: Done.

File: extensions/amp-a4a/0.1/refresh-manager.js (hunk 7 above)
Comment (glevitzky) [reply extensions/amp-a4a/0.1/refresh-manager.js]: Done.

File: extensions/amp-a4a/0.1/refresh-manager.js [hunk 12]
```diff
@@ -151,13 +150,43 @@ const managers = {};
...
+ * @param {function():boolean=} opt_predicate
+ * @return {?RefreshManager}
+ */
+export function getRefreshManager(a4a, opt_predicate) {
+  const refreshInterval =
+      getPublisherSpecifiedRefreshInterval(a4a.element, a4a.win);
+  if (!refreshInterval) {
+    return null;
+  }
+  if (opt_predicate && !opt_predicate()) {
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/refresh-manager.js]: combine refreshInterval here
